Quote parts of the table name

greg0ire opened the pull request:
In #11811, I wrongly assumed that
`$tableName` would never contain a dot as I was not able to write a test
that caused that to happen.

The secret recipe appears to be to define a schema and to quote the
table name.

Fixes https://github.com/doctrine/orm/issues/12041

greg0ire commented:
Ah no, the revert works, but for some reason, the test I'm adding has a side effect :thinking:

greg0ire commented:
Ah that's because the same schema name (`myschema`) is used in both tests.

greg0ire committed:
Quote parts of the table name

In aa141bf001588ab43e126db4152ec9b14e282f49, I wrongly assumed that
$tableName would never contain a dot as I was not able to write a test
that caused that to happen.

The secret recipe appears to be to define a schema and to quote the
table name.

To fix it for the table name, I am calling quoteSingleIdentifier()
before doing the concatenation between schema name and table name.

To fix it for the sequence name, which seems only useful when using DBAL
3 for some reason, I reuse some of the logic of the deprecated method.

Fixes #12041

greg0ire commented on src/Mapping/DefaultQuoteStrategy.php:
```diff
@@ -41,9 +44,17 @@ public function getTableName(ClassMetadata $class, AbstractPlatform $platform):
             $tableName = $class->table['schema'] . '.' . $class->table['name'];
         }
```
Just realized this is what's causing the infamous dot. Instead of inlining `quoteIdentifier()`, let me try to modify this.

morozov commented on src/Mapping/DefaultQuoteStrategy.php:
```diff
@@ -52,7 +61,10 @@ public function getTableName(ClassMetadata $class, AbstractPlatform $platform):
     {
         return isset($definition['quoted'])
-            ? $platform->quoteSingleIdentifier($definition['sequenceName'])
+            ? implode('.', array_map(
```
I have little understanding of what tradeoffs are acceptable here, so I cannot provide a reasonable feedback.

Every `implode('.', $name)` is a potential bug, which is why `quoteIdentifier()` was deprecated in the first place.

Assuming that it fixes the reported issue (which is covered by a new test), the patch looks good.

greg0ire replied:
I initially wrote a separate test just for the reported issue, and the other part of the patch fixed it. This is here to fix another issue found when I decided to reuse the existing test, if I recall correctly. In any case, I think that on a patch branch, doing this is the right thing, but that on the next minor, we might want deprecate relying on that implode. This will require further analysis as to the ramification of the deprecation though.

morozov commented on tests/Tests/ORM/Functional/Ticket/DDC2825Test.php:
```diff
@@ -42,7 +46,11 @@ public function testClassSchemaMappingsValidity(string $className, string $expec
         self::assertEquals($expectedSchemaName, $classMetadata->table['schema']);
 
         if ($this->_em->getConnection()->getDatabasePlatform()->supportsSchemas()) {
-            $fullTableName = sprintf('%s.%s', $expectedSchemaName, $expectedTableName);
+            $fullTableName = sprintf(
+                $isQuoted ? '"%s"."%s"' : '%s.%s',
+                $expectedSchemaName,
+                $expectedTableName,
+            );
         } else {
```
This kind of test looks... not really useful. On the one hand, it's an integration test that that covers implementation details, which in turn depend on the environment, so it's hard to reason whether it provides the right feedback or not. On the other, there are two levels of code branching (`if` and `?`), but a proper test shouldn't have any conditional logic.

Would it make sense to convert it to a unit test and make the expected values explicit?

Also, the dependency of the result on `supportsSchemas()` looks a bit inadequate. The result should depend only on the input (i.e. the class metadata).

greg0ire replied:
Good points. To elaborate on your last point, the test is skipped unless the database supports schemas: https://github.com/doctrine/orm/blob/0f32569a7a64b5d8e54aa2cecc10376c3be0db56/tests/Tests/ORM/Functional/Ticket/DDC2825Test.php#L28-L30

So it makes no sense to have this `if` statement: https://github.com/doctrine/orm/blob/0f32569a7a64b5d8e54aa2cecc10376c3be0db56/tests/Tests/ORM/Functional/Ticket/DDC2825Test.php#L44-L48 , since a41c6d32c3e18e7e5c008a7b2c64d62acf24966f

This test definitely needs some love, and I will start by removing this `if` statement.

greg0ire replied:
I will look into doing the improvements you mentioned, but for now, here is a PR that addresses the issue in my last message. https://github.com/doctrine/orm/pull/12094

greg0ire replied:
@morozov I did the conversion to unit tests, please take another look.

greg0ire committed:
Convert test into 2 unit tests

That test was testing too many thing and not really making it clear what
the expected output was, given some output. Instead, let us create 2
tests, each pertaining to the class under test.

greg0ire committed:
Improve comment

This comment is rendered useless by the phpdoc comment below it.
Instead, we can comment on what exactly "quoted" means.

morozov reviewed (approved):
Looks much better. Thanks!

stlrnz commented:
Hey, coming from https://github.com/doctrine/orm/issues/12041, I just wanted to say thank you for your hard work to fix the issue! I really appreciate this :)
